transport/x/cherami/config.go

```diff
@@ -0,0 +1,269 @@
+// discover Cherami Frontends if the user did not provide their own peer list.
+func DefaultPeerList(path string) TransportSpecOption {
+	return func(ts *transportSpec) {
+		ts.defaultPeerList = path
+	}
+}
+
+// TransportSpec builds a TransportSpec for the Cherami transport.
+// TransportSpecOptions may be passed to this function to configure the
+// behavior of the TransportSpec.
+//
+// 	configurator.MustRegisterTransport(
+// 		cherami.TransportSpec(cherami.DefaultPeerList("/etc/hosts.json")),
+// 	)
+//
+// See TransportConfig, InboundConfig, and OutboundConfig for details on the
+// different configuration parameters supported by this Transport.
+func TransportSpec(opts ...TransportSpecOption) config.TransportSpec {
+	var ts transportSpec
+	for _, opt := range opts {
+		opt(&ts)
+	}
+	return ts.Spec()
+}
+
+// When building inbounds and outbounds, instead of casting
+// transport.Transport to *Transport, we'll cast to the interface
+// cheramiTransport so that we can test against a mock cheramiTransport.
+type cheramiTransport interface {
+	transport.Transport
+
+	NewInbound(InboundOptions) *Inbound
+	NewOutbound(OutboundOptions) *Outbound
+}
+
+var _ cheramiTransport = (*Transport)(nil)
+
+// TransportSpec holds the configurable parts of the Cherami TransportSpec.
+type transportSpec struct {
+	defaultPeerList string
+}
+
+func (ts *transportSpec) Spec() config.TransportSpec {
+	return config.TransportSpec{
+		Name:                "cherami",
+		BuildTransport:      ts.buildTransport,
+		BuildInbound:        ts.buildInbound,
+		BuildOnewayOutbound: ts.buildOnewayOutbound,
+	}
+}
+
+// TransportConfig configures the shared Cherami Transport. This is shared
+// between all Cherami outbounds and inbounds of a Dispatcher.
+//
+// All fields in TransportConfig are optional and this section may be skipped
+// entirely for most use cases.
+type TransportConfig struct {
+	// If specified, this is the address of a specific Cherami Frontend or a
+	// locally hosted development instance. Most users should omit this since
+	// the Cherami Frontend will be discovered automatically.
+	//
+	// 	address: 127.0.0.1:4922
+	Address string `config:"address"`
+
+	// Path to a JSON file containing the TChannel peer list used to
+	// auto-discover Cherami Frontend machines. This may be skipped if a
+	// default peer list was provided on TransportSpec instantiation or if an
+	// Address was provided.
+	//
+	// 	peerList: /etc/hosts.json
+	PeerList string `config:"peerList"`
+
+	// Timeout for requests to the Cherami service. The default timeout should
+	// suffice for most use cases.
+	//
+	// 	timeout: 5s
+	Timeout time.Duration `config:"timeout"`
+
+	// Name of the Cherami deployment to which requests should be sent. Some
+	// valid values are, "prod", "staging", "staging2", and "dev".
+	//
+	// 	deploymentStr: dev
+	//
+	// Defaults to "prod".
+	DeploymentStr string `config:"deploymentStr"`
```

Comment:
I don't like the `Str` suffix, this isn't like anything we do in other areas. It would be nice if `DeploymentType` was an enum, ie `type DeploymentType int` and we could parse this from a string.